updatehub: report error before rollback on unconfirmed image

adrienmaillard03 opened the pull request:
# Summary
Fixes an issue where unconfirmed images are rolled back without notifying the UpdateHub server.

# Motivation
When using updatehub_autohandler() with CONFIG_BOOT_UPGRADE_ONLY=n, unconfirmed images are correctly rolled back. However, the UpdateHub agent does not send an "error" report to the server, causing it to re-offer the same broken update.

There is the same problem when using manual mode, there are no ways to report the failure to the Updatehub server.

# Fix
Sends an error report before rebooting on rollback in updatehub_autohandler().
Expose a updatehub_report_error() function, that allows manual mode to report an error to the server.

Commit:
updatehub: report error before rollback on unconfirmed image

Report the error state to the UpdateHub server before triggering a rollback
in autohandler mode. This prevents the server from redeploying the same
(failed) update after rollback, avoiding update loops. Also exposes a new
public syscall: updatehub_report_error(), allowing manual mode users to
report an error explicitly.

Signed-off-by: Adrien Maillard <adrien.maillard@edu.hefr.ch>

nandojve commented:
CC: @otavio

adrienmaillard03 commented:
This PR fixes the missing error report in automatic mode (updatehub_autohandler()), but does not address the same issue in manual mode.
Currently, there's no way to report errors or deployment status when using the manual flow, as the internal report() function is not exposed.

adrienmaillard03 commented:
It is now possible to report a failure in manual mode :
```c
  switch (updatehub_probe())
  {
  case UPDATEHUB_UNCONFIRMED_IMAGE:
      LOG_ERR("Image is unconfirmed. Rebooting to revert back to previous");
      updatehub_report_error(); // report error to the server
      updatehub_reboot();
      break;
  case UPDATEHUB_HAS_UPDATE:
      switch (updatehub_update())
      {
      case UPDATEHUB_OK:
          updatehub_reboot();
          break;

      default:
          LOG_ERR("Error installing update.");
          break;
      }
  }
```

otavio commented on subsys/mgmt/updatehub/updatehub.c:
```diff
@@ -1001,7 +1001,9 @@ static void autohandler(struct k_work *work)
 		LOG_ERR("Image is unconfirmed. Rebooting to revert back to previous"
 			"confirmed image.");
-
+		if (report(UPDATEHUB_STATE_ERROR) < 0) {
```
You could have worked this method call to use the new API method.

adrienmaillard03 replied:
I have change my implementation

Commit:
updatehub: use the new API method in autohandler

Call updatehub_report_error() when detecting an unconfirmed image in
autohandler to ensure the server is notified of the failure before
triggering a rollback.

Signed-off-by: Adrien Maillard <adrien.maillard@edu.hefr.ch>

kartben commented on include/zephyr/mgmt/updatehub.h:
```diff
@@ -90,6 +90,15 @@ __syscall int updatehub_confirm(void);
+ *
+ * @return 0 on success, negative errno on failure.
+ */
+__syscall int updatehub_report_error(void);
```
Please add an entry for this new API in the 4.2 release notes

kartben replied:
@nandojve I just added it...

nandojve reviewed (dismissed):
Thank you @kartben !

kartben commented:
rebased due to conflict in release notess

kartben commented:
I am taking the liberty to approve as the change I made to the branch is trivial (adding entry to release notes) and shouldn't prevent me from giving a +1 for the rest :)
